intellij-plugin/src/main/kotlin/com/apollographql/ijplugin/gradle/GradleToolingModelService.kt

```diff
@@ -0,0 +1,161 @@
+import com.intellij.openapi.vfs.VfsUtilCore
+import com.intellij.openapi.vfs.VirtualFileManager
+import org.gradle.tooling.CancellationTokenSource
+import org.gradle.tooling.GradleConnector
+import org.gradle.tooling.model.GradleProject
+import org.jetbrains.kotlin.idea.framework.GRADLE_SYSTEM_ID
+import org.jetbrains.plugins.gradle.service.execution.GradleExecutionHelper
+import org.jetbrains.plugins.gradle.settings.GradleExecutionSettings
+import org.jetbrains.plugins.gradle.util.GradleConstants
+import java.io.File
+import java.util.concurrent.Executors
+
+class GradleToolingModelService(
+    private val project: Project,
+) : Disposable {
+  private var gradleCancellation: CancellationTokenSource? = null
+  private var abortRequested: Boolean = false
+  private val gradleExecutorService = Executors.newSingleThreadExecutor()
+
+  init {
+    logd("project=${project.name}")
+    startObserveGradleHasSynced()
+  }
+
+  private fun startObserveGradleHasSynced() {
+    logd()
+    project.messageBus.connect(this).subscribe(GradleHasSyncedListener.TOPIC, object : GradleHasSyncedListener {
+      override fun gradleHasSynced() {
+        logd()
+        abortFetchToolingModels()
+
+        if (project.apolloProjectService.isApolloKotlin3Project) {
+          fetchToolingModels()
+        }
+      }
+    })
+  }
+
+  private fun fetchToolingModels() {
+    logd()
+
+    if (gradleCancellation != null) {
+      logd("Already running")
+      return
+    }
+
+    val rootProjectPath = project.getGradleRootPath() ?: return
+    gradleExecutorService.submit {
+      abortRequested = false
+      val gradleExecutionHelper = GradleExecutionHelper()
+      val executionSettings = ExternalSystemApiUtil.getExecutionSettings<GradleExecutionSettings>(project, rootProjectPath, GradleConstants.SYSTEM_ID)
+      val rootGradleProject = gradleExecutionHelper.execute(rootProjectPath, executionSettings) { connection ->
+        gradleCancellation = GradleConnector.newCancellationTokenSource()
+        logd("Fetch Gradle project model")
+        return@execute try {
+          val id = ExternalSystemTaskId.create(GRADLE_SYSTEM_ID, ExternalSystemTaskType.RESOLVE_PROJECT, project)
+          gradleExecutionHelper.getModelBuilder(GradleProject::class.java, id, executionSettings, connection, ExternalSystemTaskNotificationListenerAdapter.NULL_OBJECT)
+              .withCancellationToken(gradleCancellation!!.token())
+              .get()
+        } catch (t: Throwable) {
+          logw(t, "Couldn't fetch Gradle project model")
+          null
+        } finally {
+          gradleCancellation = null
+        }
+      } ?: return@submit
+
+      // We're only interested in projects that apply the Apollo plugin - and thus have the codegen task registered
+      val allApolloGradleProjects: List<GradleProject> = (rootGradleProject.children + rootGradleProject)
+          .filter { it.tasks.any { it.name == CODEGEN_GRADLE_TASK_NAME } }
```

Comment:
This creates coupling in the unlikely event that we rename "generateApolloSources". Could you ask the model for every project and "just" fail on those that do not have the Apollo plugin installed?

Reply:
The issue is that we iterate on subprojects to get the model, and if you have a project with a huge amount of modules, that will be impractical (and unneeded if most of them are not Apollo).
It's a bit unfortunate that there's no way to know what plugins are applied from `GradleProject` directly.

Agreed that the coupling to the task name would best be avoided though.

Reply:
>  that will be impractical 

impractical as in "slow"?

Reply:
As in "so slow that it's unusable" :) (I'm thinking of projects with hundreds or even thousands of modules).

Reply:
👍